File: src/main/java/org/antechrestos/springboot/tracingissue/infrastructure/MongoRepository.java
package org.antechrestos.springboot.tracingissue.infrastructure;

import org.antechrestos.springboot.tracingissue.domain.User;
import org.antechrestos.springboot.tracingissue.domain.UserRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Repository;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Repository
class MongoRepository implements UserRepository {

    private static final Logger logger = LoggerFactory.getLogger(MongoRepository.class);
    private final UserRepositoryMongoStore mongoStore;

    MongoRepository(UserRepositoryMongoStore mongoStore) {
        this.mongoStore = mongoStore;
    }


    @Override
    public Mono<User> create(User user) {
        return mongoStore.save(new DBUser(user.id(), user.firstname(), user.name()))
                .map(this::toDomain)
                .doOnEach(userSignal -> logger.debug("Saved {}", userSignal.get()));
    }

    @Override
    public Flux<User> list() {
        return mongoStore.findAll().map(this::toDomain);
    }

    @Override
    public Mono<User> get(String id) {
        return mongoStore.findById(id)
                .map(this::toDomain)
                .doOnEach(userSignal -> logger.debug("Saved {}", userSignal.get()));
    }

    @Override
    public Mono<Void> delete(String id) {
        return mongoStore.deleteById(id);
    }

    private User toDomain(DBUser dbUser) {
        return new User(dbUser.id(), dbUser.firstname(), dbUser.name());
    }
}

File: src/main/kotlin/org/antechrestos/springboot/tracingissue/infrastructure/MongoRepository.kt
package org.antechrestos.springboot.tracingissue.infrastructure

import kotlinx.coroutines.flow.map
import org.antechrestos.springboot.tracingissue.domain.User
import org.antechrestos.springboot.tracingissue.domain.UserRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Repository

@Repository
class MongoRepository(private val mongoStore: UserRepositoryMongoStore) : UserRepository {


    private val logger = LoggerFactory.getLogger(this.javaClass)

    override suspend fun create(user: User) =
        mongoStore.save(
            user.also { logger.debug("about to create {}", it) }
                .let {
                    DBUser(
                        id = it.id,
                        firstname = it.firstname,
                        name = it.name
                    )
                }
        )
            .let { toDomain(it) }
            .also { logger.debug("user created {}", it) }

    override fun list() = mongoStore.findAll().map { toDomain(it) }


    override suspend fun get(id: String) = mongoStore.findById(id.also { logger.debug("loading with id {}", it) })
        ?.let { toDomain(it) }
        ?.also {
            logger.debug("user retrieved {}", it)
        }

    override suspend fun delete(id: String) {
        mongoStore.deleteById(id.also { logger.debug("deleting with id {}", id) })
            .also { logger.debug("user deleted {}", id) }

    }

    private fun toDomain(it: DBUser) = User(
        id = it.id,
        firstname = it.firstname,
        name = it.name
    )
}